docs/proposals/pod-troubleshooting.md

```diff
@@ -0,0 +1,489 @@
+mechanism to execute a shell or other troubleshooting tools inside a running pod
+without requiring that the associated container images include such tools.
+
+## Motivation
+
+Many developers of native Kubernetes applications wish to treat Kubernetes as an
+execution platform for custom binaries produced by a build system. These users
+can forgo the scripted OS install of traditional Dockerfiles and instead `COPY`
+the output of their build system into a container image built `FROM scratch`.
+This confers several advantages:
+
+1.  **Minimal images** lower operational burden and reduce attack vectors.
+1.  **Immutable images** improve correctness and reliability.
+1.  **Smaller image size** reduces resource usage and speeds deployments.
+
+The disadvantage of using containers built `FROM scratch` is the lack of system
+binaries provided a base Operation System image makes it difficult to
+troubleshoot running containers. Kubernetes should enable troubleshooting pods
+regardless of the contents of the container image.
+
+## Goals and Non-Goals
+
+Goals include:
+
+*   Enable troubleshooting of minimal container images
+*   Allow troubleshooting of containers in `CrashLoopBackoff`
+*   Improve supportability of native Kubernetes applications
+*   Enable novel uses of short-lived containers in pods
+
+Non-Goals of this proposal are:
+
+*   Fully design the ability to cross-mount container image filesystems. This is
+    complex and should be covered in a separate doc.
+*   Guarantee resources for ad-hoc troubleshooting. If troubleshooting causes a
+    pod to exceed its resource limit it may be evicted.
+
+## Requirements
+
+A solution to troubleshoot minimal container images MUST:
+
+*   troubleshoot arbitrary running containers with minimal prior configuration
+*   allow access to all namespaces shared by containers in a pod and the mount
+    namespaces of individual containers
+*   fetch troubleshooting utilities at debug time rather than at the time of pod
+    initialization
+*   respect admission restrictions
+*   allow introspection of pod state using existing tools (no hidden containers)
+*   support arbitrary runtimes via the CRI (possibly with reduced feature set)
+
+A good solution SHOULD:
+
+*   have an excellent user experience (i.e. should be a feature of the platform
+    rather than a config-time solution)
+*   require no direct access to the node
+*   have no inherent side effects to the running container image
+*   re-use existing container image distribution channels
+*   enable detection of pods that have been modified by an exec ("tainted" pods)
+
+## kubectl debug
+
+We will introduce an "execute a debug container in a pod" pattern that parallels
+`kubectl exec`'s "execute a debug process in a container" in a new command,
+`kubectl debug`, which has two modes of operation:
+
+1.  *Copy Debug* mode creates a copy of a pod with minor changes to the pod spec
```

Comment:
I think the doc describes use cases in sufficient detail. What are your reasons for not wanting to add this feature as described? It's hard to argue with feelings …

That said, I tend to agree that *debug* is not the best subcommand / verb for the *Copy Debug* mode because it is not discoverable / intuitive: reading through the list of available subcommands I would not expect it to do what it does. *kubectl clone pod* might be better in that respect and could eventually be extended to work on other types of resources as well.